Add action for launching installed applications

jaimeMF opened the pull request:
The applications use the SailJail sandbox if they are configured to do so.

Resolves https://github.com/sailfishos-chum/sailfishos-chum-gui/issues/105

Olf0 commented:
@rinigus and @piggz, to properly review and test this while still fighting the flu is not going to work at all, because reviewing this is at the boarder of by capabilities, any way.  But if one of you reviews / tests and signals "O.K." (with or without introducing changes or requesting ones from jamieMF), I will create a release v0.6.7.

Binaries to test are here: https://github.com/sailfishos-chum/sailfishos-chum-gui/actions/runs/7258613885

piggz commented on src/chumpackage.cpp:
```diff
@@ -254,6 +254,21 @@ void ChumPackage::setInstalledVersion(const QString &v)
     emit updated(m_id, PackageInstalledRole);
+    if (type() == PackageApplicationDesktop && installed()) {
+        auto trfl = Daemon::getFiles(m_pkid_installed);
+        connect(trfl, &Transaction::files, this, [this](
```
Is there a reason for capturing [this] in the lambda, it appears unused?

jaimeMF replied:
If I just put `[]` it complains because it tries to access `m_desktopFile` , `updated` and `m_id`. By using `[=]` o `[&]` it also compiles. Since I have a limited knowledge of C++  I would need to read with some attention the documentation to grasp the differences.

Olf0 replied:
@piggz, in my perception @jaimeMF's reply shows, that his C++ skills are not that far beyond mine: Apparently he tried things until it was compiling and working, hence asking for proper semantics, plus reasoning for them is futile.

As you do not seem to oppose this implementation to address issue #105 fundamentally, why don't you make a suggestion or MR on top of this with a semantically proper lambda expression.  Because I am afraid without such a move this MR will stall forever.

jaimeMF replied:
> @piggz, in my perception @jaimeMF's reply shows, that his C++ skills are not that far beyond mine: Apparently he tried things until it was compiling and working, hence asking for proper semantics, plus reasoning for them is futile.
> 
> As you do not seem to oppose this implementation to address issue #105 fundamentally, why don't you make a suggestion or MR on top of this with a semantically proper lambda expression. Because I am afraid without such a move this MR will stall forever.

Indeed, my C++ skills are quite limited, so feel free to suggest any modification to improve it.

Olf0 replied:
@piggz, you seem to have had a change (= enhancement) of the proposed code in mind, please specify which.

Olf0 replied:
@piggz, I do not want to stall @jaimeMF's effort forever, hence please speak up or I will merge this in April, because it seems to be working fine and it can always still be improved later, because it is just software.

jaimeMF replied:
In the file chum.cpp there are multiple cases where `this` is captured in the lambdas, I think I used it as a reference. Would it make it clearer if i used `this->` to access the the fields inside the lambda?

Olf0 commented:
@jaimeMF, as stated, I am willing to merge this, because all requests to the original authors to review this have been futile.

But lastly now I have a (maybe silly) [question](https://github.com/sailfishos-chum/sailfishos-chum-gui/pull/266/files/1899839c30c24c927c7df477ee787227bad5138f), I would like to have clarified, first.

Olf0 commented:
@nephros, you would do the progress of the SailfishOS:Chum GUI app a big favour, if you add concise comments to the two open review comment threads above.  I did what I can, but neither @piggz or @jaimeMF respond any longer, this is why I am resorting to you; sometimes (more often lately) it appears that we are the only ones left who care about community-related infrastructure software for SailfishOS. :disappointed:

jaimeMF commented:
> > On the topic of design: I'm not convinced parsing the desktop Exec line is the ideal way to do this. At the least it's fragile should the sailjail parameters change in some way. I also believe this will probably fail in some esoteric cases.
> > I would prefer if there was some way to let Lipstick handle all of this, i.e. just simulate a user click.
> > One way I have found is
> > ```
> >  busctl --user call org.nemomobile.lipstick / org.sailfishos.fileservice openUrl s file:///usr/share/applications/foo.desktop
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > buut this launching sn application may just be coincidence, not too sure.
> 
> ... So something like this:
> 
> [nephros@689a26f](https://github.com/nephros/sailfishos-chum-gui/commit/689a26fad9884be4f80941e7e0057252bbe67351)

I tested your version and it seems to behave like mine, so  we could use it as it is indeed simpler and hopefully more robust.

Olf0 commented:
Thank you very much, @nephros for taking a look at this PR, then coming up, researching and implementing a simpler solution, and @jaimeMF for the original idea and testing @nephros' variant.

Hence I create PR #274: Is that what you @nephros had in mind.  Is the other commit (rsp. its changes) in that [branch `applaunch`](https://github.com/nephros/sailfishos-chum-gui/tree/applaunch) good to be merged, too?

Olf0 commented:
> i believe it's not not caring, it's about resources being spread thin.

Actually IMO this is basically the same: A matter of prioritising.  More on this below.

> Certainly in my case, way too many irons in the forge as they say.

This sure applies to piggz, rinigus an me, too.  Still, you did help here and I try to handle PRs and issues properly (just the handling).

> And in the light of this I don't think it's smart lashing out at contributors this way. It's certainly not helping.

1. Absolutely agreed, that grumpy statements without context and real message (besides complaining) are never helpful.
2. "Contributors" is not correct: They are the two creators and maintainers remaining from the original trio, after mentaljam (Petr) ceased working FOSS@GitHub.  I became co-maintainer, because I offered to manage releases and maintain the infrastructure (CI workflows etc.), but I always clearly emphasised that I cannot perform coding or proper code review beyond shell and simple C code.  In my understanding we had a kind of informal agreement and my idea was that I free the two from the tasks I can perform well in order to let them do what they are best at: Care about code, here and their other projects (e.g. community ports of SFOS, Pure Maps etc.).  This is why I do feel left behind as I apparently have become the only maintainer of the SailfishOS:Chum GUI app, because almost all of my requests to comment on or review code have been met with silence.
3. I think it is unfair to contributors as Jamie, you and others not to react to PRs and it would be detrimental to the development to not trying to integrate any useful contribution.  Specifically here it was not nice at all to start a code review, request changes from Jamie and then turn silent, formally stalling the whole PR with no chance for me to resolve the situation properly without taking a radical shortcut as either declining and closing the PR or to merge it blindly.
4. Yes I am very frustrated, because I spend a good part of my spare time with FLOSS which are not in my primary interest and stall the development of things I want to implement, because a feeling of responsibility and aforementioned social aspects keep me busy with administrative tasks and code reviews which are actually beyond my capabilities, but I know no other person will perform them.

----------------------------------

> On the matter of reviewing, I don't think I'm a good choice for CPP. I can write things that compile, sometimes. But I don't really _know_ Qt or cpp.

LOL! :wink:
Your code is well designed and well readable even for me, and it does not just compile but also work well in practice, as I can tell from years of experience and collaboration. :smiley:

> I shall do my best though.

You always do, from what I have observed. :smile:

jaimeMF commented:
> > > > On the topic of design: I'm not convinced parsing the desktop Exec line is the ideal way to do this. At the least it's fragile should the sailjail parameters change in some way. I also believe this will probably fail in some esoteric cases.
> > > > I would prefer if there was some way to let Lipstick handle all of this, i.e. just simulate a user click.
> > > > One way I have found is
> > > > ```
> > > >  busctl --user call org.nemomobile.lipstick / org.sailfishos.fileservice openUrl s file:///usr/share/applications/foo.desktop
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > buut this launching sn application may just be coincidence, not too sure.
> > > 
> > > 
> > > ... So something like this:
> > > [nephros@689a26f](https://github.com/nephros/sailfishos-chum-gui/commit/689a26fad9884be4f80941e7e0057252bbe67351)
> > 
> > 
> > I tested your version and it seems to behave like mine, so we could use it as it is indeed simpler and hopefully more robust.
> 
> Nice! So, will you add this change to the PR branch, or shall I do it and put a PR onto your branch?

@Olf0 already opened a PR to the main repo with your changes. If you prefer that I use my PR instead, I can add your changes.

Olf0 commented:
> @Olf0 already opened a PR to the main repo with your changes. If you prefer that I use my PR instead, I can add your changes.

See PR #274, which applies cleanly in contrast to this one (likely due to changes in the `main` branch meanwhile).

But ultimately I do not care at all which route is taken, I just wanted to make it easy for both of you to look at it (~ review) and give an "O.K." so I can merge.

jaimeMF commented:
I have rebased on top of main and cherry picked @nephros 's commit.

Olf0 commented:
> I have rebased on top of main and cherry picked @nephros 's commit.

Than you very much @jaimeMF: This is really looking good now!

Still, I would really appreciate if you provide me with your opinion on [my code review and comment](https://github.com/sailfishos-chum/sailfishos-chum-gui/pull/266#pullrequestreview-1956473484): In the light of the recent changes (the ones suggested by nephros, i.e. not to evaluate the desktop-file-entry for starting the app, but to use the launcher), I tend to my second suggestion.But as noted before, that may be nonsense (as anything I write), then I would like to understand the reason.

jaimeMF commented:
> > I have rebased on top of main and cherry picked @nephros 's commit.
> 
> Than you very much @jaimeMF: This is really looking good now!
> 
> Still, I would really appreciate if you provide me with your opinion on [my code review and comment](https://github.com/sailfishos-chum/sailfishos-chum-gui/pull/266#pullrequestreview-1956473484): In the light of the recent changes (the ones suggested by nephros, i.e. not to evaluate the desktop-file-entry's `exec=` line for starting the app, but to use the launcher), I tend to my second suggestion or simply `visible: (pkg.installed && pkg.desktopFile.length > 0)`. But as noted before, that may be nonsense (as anything I write), then I would like to understand the reason.

The `pkg.type == ChumPackage.PackageApplicationDesktop` check seems unnecessary, I have removed it. Thank you.

Olf0 commented on qml/pages/PackagePage.qml:
```diff
@@ -54,6 +54,12 @@ Page {
+            MenuItem {
+                //% "Launch"
+                text: qsTrId("chum-launch")
+                visible: pkg.installed && pkg.type == ChumPackage.PackageApplicationDesktop && pkg.desktopFile.length > 0
```
Reply is at https://github.com/sailfishos-chum/sailfishos-chum-gui/pull/266#issuecomment-2113492583.

Olf0 reviewed (approved):
LGTM (AFAICT), so let's go!

Olf0 commented:
SailfishOS-Chum GUI application v0.6.7 is available at the [`sailfishos:chum:testing`](https://build.sailfishos.org/package/show/sailfishos:chum:testing/sailfishos-chum-gui) repository for testing.
I will submit it to the `sailfishos:chum:testing` repo proper in a couple of days, if no negative feedback is provided.

Olf0 commented:
Oh, I missed that there is a new translatable string, hence did not run `lupdate` and updated the translations at Transifex.

Maybe this weekend.

jaimeMF commented:
> Second thing is, and I'm sorry I didn't catch that earlier:
> 
> Launched applications show up as not-sailjailed! They seem to inherit the out-of-jail status of Chum GUI.

Where do they show up as not-sailjailed? When launching a newly installed application it ask for the permissions.

Olf0 commented:
BTW, the new menu entry "Launch" does not show on SailfishOS 3.2.1.  This is not tragic, as it is just not there, but mind that SailfishOS:Chum (the repository and all infrastructure, including the GUI app) does support all SailfishOS releases ≥ 3.1.0.

Olf0 commented:
> Oh, I missed that there is a new translatable string, hence did not run `lupdate` and updated the translations at Transifex.

Done by commit f54ae3a.

Olf0 commented:
Oh, ugh, launching from the console does not look good (part 1: git-hoster access):

```
[nemo@sailfishos ~]$ sailfishos-chum-gui 
[D] unknown:0 - Using Wayland-EGL
[W] unknown:0 - Failed to refresh Chum repository "Failed to obtain authentication."
[D] unknown:0 - GitLab support added for "gitlab.com"
[D] unknown:0 - Forgejo support added for "codeberg.org"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "flypig/harbour-fosdem23"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/harbour-privoxy"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/nextcloud-client"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/harbour-uptimed"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/openrepos-keychain"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - Forgejo: Failed to fetch repository data for Forgejo "sailfish-moreutils"
[W] unknown:0 - Forgejo: Error:  "Error transferring https://codeberg.org/api/v1/repos/sailfish-moreutils - server replied: Not Found"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "project/nephros/coderus-screencast"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/harbour-privoxy"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/openrepos-nethack"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/nextcloud-client"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/harbour-privoxy"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/qtkeychain"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/harbour-fbjail"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/harbour-mlocate"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "dracks/quicklaunchslx"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/youtube-dl/-/tree/sfos-open-helper"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/harbour-privoxy"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "nephros/openrepos-nethack"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - Forgejo: Failed to fetch repository data for Forgejo "sailfish-moreutils"
[W] unknown:0 - Forgejo: Error:  "Error transferring https://codeberg.org/api/v1/repos/sailfish-moreutils - server replied: Not Found"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "sailfishos-chum/smem"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
[W] unknown:0 - GitLab: Failed to fetch repository data for GitLab "dracks/flight-tracker-slx"
[W] unknown:0 - GitLab: Error:  "Host requires authentication"
```

Do you see these errors accessing GitLab, Forgejo (codeberg.org) and Sailfish-OBS (SailfishOS:Chum repository) as well?

Olf0 commented:
When I switch to ***Installed packages***, it outputs:
```
[W] unknown:17 - file:///usr/share/sailfishos-chum-gui/qml/components/PackagesListItem.qml:17:5: QML Image: Error transferring https://github.com/ilpianista/harbour-Base64/-/raw/master/icons/harbour-base64.svg - server replied: Not Found
[W] unknown:17 - file:///usr/share/sailfishos-chum-gui/qml/components/PackagesListItem.qml:17:5: QML Image: Error transferring https://github.com/black-sheep-dev/harbour-kasa/raw/master/icons/harbour-audiothek.svg - server replied: Not Found
```
O.K., well plausible so far.

Then I scrolled to ***Bar Code*** and opened it:
```
[W] unknown:157 - file:///usr/lib/qt5/qml/Sailfish/Silica/private/RemorseBase.qml:157:5: QML Column: Cannot anchor to an item that isn't a parent or sibling.
[W] unknown:157 - file:///usr/lib/qt5/qml/Sailfish/Silica/private/RemorseBase.qml:157:5: QML Column: Cannot anchor to an item that isn't a parent or sibling.
[W] unknown:157 - file:///usr/lib/qt5/qml/Sailfish/Silica/private/RemorseBase.qml:157:5: QML Column: Cannot anchor to an item that isn't a parent or sibling.
[W] unknown:157 - file:///usr/lib/qt5/qml/Sailfish/Silica/private/RemorseBase.qml:157:5: QML Column: Cannot anchor to an item that isn't a parent or sibling.
[W] unknown:157 - file:///usr/lib/qt5/qml/Sailfish/Silica/private/RemorseBase.qml:157:5: QML Column: Cannot anchor to an item that isn't a parent or sibling.
[D] unknown:0 - bool PatchManagerTranslator::installTranslator(const QString&) "return-old-remorse-animation" QLocale(C, Default, Default)
[D] unknown:0 - bool PatchManagerTranslator::installTranslator(const QString&) fail
[D] unknown:0 - bool PatchManagerTranslator::installTranslator(const QString&) "return-old-remorse-animation" QLocale(C, Default, Default)
[D] unknown:0 - bool PatchManagerTranslator::installTranslator(const QString&) fail
```
I guess, I better switch off the Patchmanager-Patch ***Return old remorse item animation***.

Oh, it is not enabled (because it is not applicable on SailfishOS 3.2.1) and it tries to patch `/usr/lib/qt5/qml/Sailfish/Silica/RemorseItem.qml:204&236` (i.e. not `/usr/lib/qt5/qml/Sailfish/Silica/private/RemorseBase.qml:157`)!?!  Ah, but ***Return old remorse popup animation*** patches `/usr/lib/qt5/qml/Sailfish/Silica/private/RemorseBase.qml`: Disabled and deactivated it.

Next try, this time I also watched the GUI:
1. WRT part 1: Despite `[W] unknown:0 - Failed to refresh Chum repository "Failed to obtain authentication."` being emitted again in the terminal window, the GUI output suggests that refreshing of the SailfishOS:Chum  repository works fine.
2. When entering ***Installed packages*** the two `QML image` errors are displayed as expected, but neither when entering ***Bar Code** or ***Bugger*** any additional command line output appears.  Well, but there still is no pulley menu entry ***Start application*** (`chum-launch`), only ***Remove***, plus ***Source Code***, ***Issue tracker*** and ***Discussion forum*** for ***BUgger***.

-------------

#### TL;DR
No, nothing in terminal output of the SailfishOS:Chum GUI app started at the command line indicates anything WRT the pulley entry `chum-launch`, but this entry (with its text ***Start application***) is not shown on SailfishOS 3.2.1.

Olf0 commented:
> > > Second thing is, and I'm sorry I didn't catch that earlier:
> > > Launched applications show up as not-sailjailed! They seem to inherit the out-of-jail status of Chum GUI.
> > 
> > 
> > Where do they show up as not-sailjailed? When launching a newly installed application, it asks for permissions.
> 
> i'll have to check in detail then.
> 
> I'm just using my Sandbox Indicator patch to check, it's possible that that is wrong.

Any news WRT this observation, @nephros?

Olf0 commented:
The SailfishOS:Chum GUI app v0.6.7-2 is available at [its GitHub releases page](https://github.com/sailfishos-chum/sailfishos-chum-gui/releases) and in [the `sailfishos:chum:testing` repository](https://build.sailfishos.org/package/show/sailfishos:chum:testing/sailfishos-chum-gui): Please check that it works fine, especially WRT [the aforementioned two points](https://github.com/sailfishos-chum/sailfishos-chum-gui/pull/266#issuecomment-2128004199).

Olf0 commented:
> Is it possible the github/lab/codeberg errors are because of older SSL certificates or library versions in the testing OS?

Unlikely, because I specifically fixed-up these things and everything else works, e.g. HTTPS for websites in Browser.

But this exactly is the reason, why I asked you and @jaimeMF to test.

Olf0 commented:
O.K., I filed issue #289 WRT the command line output, i.e. the warnings indicating authentication failures at the Sailfish-OBS, GitLab.com and codeberg.org.

------------------

Any news WRT the apparent sandboxing issues?
I.e. [was this a false positive](https://github.com/sailfishos-chum/sailfishos-chum-gui/pull/266#issuecomment-2115595163) or is it a real issue?
